File: be/src/main/java/sg/flow/controllers/UserController.java
package sg.flow.controllers;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;

import sg.flow.models.user.UpdateUserProfile;
import sg.flow.models.user.UserProfile;
import sg.flow.services.UserServices.UserService;

@Controller
public class UserController {

    private final String USER = "/user";

    @Autowired
    private UserService userService;

    @GetMapping(USER + "/getUserProfile")
    public ResponseEntity<UserProfile> getUserProfile(@AuthenticationPrincipal(expression = "userId") Integer userId) {
        return ResponseEntity.ok(userService.getUserProfile(userId));
    }

    @GetMapping(USER + "/getUserPreferenceJson")
    public ResponseEntity<String> getUserPreferenceJson(
            @AuthenticationPrincipal(expression = "userId") Integer userId) {
        return ResponseEntity.ok(userService.getUserPreferenceJson(userId));
    }

    @PostMapping(USER + "/updateUserProfile")
    public ResponseEntity<UserProfile> updateUserProfile(@AuthenticationPrincipal(expression = "userId") Integer userId,
            @RequestBody UpdateUserProfile userProfile) {
        return ResponseEntity.ok(userService.updateUserProfile(userId, userProfile));
    }
}

File: be/src/main/kotlin/sg/flow/controllers/UserController.kt
package sg.flow.controllers

import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController
import sg.flow.models.user.UpdateUserProfile
import sg.flow.models.user.UserProfile
import sg.flow.services.UserServices.UserService

@RestController
class UserController(private val userService: UserService) {

    @GetMapping("/user/getUserProfile")
    suspend fun getUserProfile(
            @AuthenticationPrincipal(expression = "userId") userId: Int
    ): ResponseEntity<UserProfile> = ResponseEntity.ok(userService.getUserProfile(userId))

    @GetMapping("/user/getUserPreferenceJson")
    suspend fun getUserPreferenceJson(
            @AuthenticationPrincipal(expression = "userId") userId: Int
    ): ResponseEntity<String> = ResponseEntity.ok(userService.getUserPreferenceJson(userId))

    @PostMapping("/user/updateUserProfile")
    suspend fun updateUserProfile(
            @AuthenticationPrincipal(expression = "userId") userId: Int,
            @RequestBody userProfile: UpdateUserProfile
    ): ResponseEntity<UserProfile> =
            ResponseEntity.ok(userService.updateUserProfile(userId, userProfile))
}
